### PR TITLE
Feature/runtime raw socket disable

### PR DESCRIPTION
**Pull request by morantr**
If raw socket fails to be opened due to insufficient permissions continue without it.

### CODE REVIEW

File: agent/src/beerocks/monitor/monitor_rssi.cpp [hunk 1]
```diff
@@ -38,8 +38,6 @@ void monitor_rssi::stop()
         close(arp_socket);
         arp_socket = -1;
     }
-    mon_db       = nullptr;
```
Comment (tomereli) [inline agent/src/beerocks/monitor/monitor_rssi.cpp:41]: Why did you delete these lines? It is not clear from the commit message.

File: agent/src/beerocks/monitor/monitor_rssi.cpp [hunk 2]
```diff
@@ -241,7 +247,7 @@ void monitor_rssi::process()
                                << " delta_val=" << int(delta_val);
                 }
             }
-            if (!conf_disable_initiative_arp) {
+            if (arp_enabled() || !conf_disable_initiative_arp) {
```
Comment (tomereli) [inline agent/src/beerocks/monitor/monitor_rssi.cpp]: Shouldn't this be `if (arp_enabled() && !conf_disable_initiative_arp)`?
And coming to think about it, shouldn't arp_enabled() be enough? i.e. move the conf_disable_initiative_arp check to inside the arp_enabled() API.

File: agent/src/beerocks/monitor/monitor_rssi.h [hunk 3]
```diff
@@ -43,6 +43,8 @@ class monitor_rssi {
     void send_rssi_measurement_response(std::string &sta_mac, monitor_sta_node *sta_node);
     void monitor_idle_station(std::string &sta_mac, monitor_sta_node *sta_node);
 
+    bool arp_enabled() { return arp_socket > 0; }
```
Comment (tomereli) [inline agent/src/beerocks/monitor/monitor_rssi.h:46]: Here also add the initiative_arp config check.

File: agent/src/beerocks/monitor/monitor_thread.cpp [hunk 4]
```diff
@@ -273,10 +273,10 @@ void monitor_thread::after_select(bool timeout)
 
             LOG(TRACE) << "mon_rssi.start()";
             if (!mon_rssi.start(&mon_db, slave_socket)) {
-                LOG(ERROR) << "mon_rssi.start() failed";
-                thread_last_error_code = MONITOR_THREAD_ERROR_ATTACH_FAIL;
-                stop_monitor_thread();
-                return;
+                // If monitor rssi failed to start, continue without it. It might failed due to
+                // insufficient permissions. Detailed error message is printed inside.
+                LOG(WARNING) << "mon_rssi.start() failed, ignore and continue without it";
```
Comment (tomereli) [inline agent/src/beerocks/monitor/monitor_thread.cpp:278]: We should log the failure reason - i.e. if its because of missing permissions or some other failure. This has to be perfectly clear from this log.

File: agent/src/beerocks/slave/backhaul_manager/wan_monitor.cpp [hunk 5]
```diff
@@ -33,7 +33,7 @@ static int netlink_open_socket()
     // Open a netlink socket
     int fd = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
     if (fd < 0) {
-        LOG(ERROR) << "Failed creating netlink socket";
+        LOG(ERROR) << "Failed creating netlink socket: " << strerror(errno);
```
Comment (tomereli) [inline agent/src/beerocks/slave/backhaul_manager/wan_monitor.cpp:36]: Good - will this print that there are no permissions to create the socket?

File: framework/platform/bpl/uci/arp/monitor/arp_monitor.cpp [hunk 6]
```diff
@@ -217,21 +219,21 @@ bool arp_monitor::start(std::string strIface)
     // open netlink sock
     if ((m_fdMonSocket = socket(AF_NETLINK, SOCK_RAW | SOCK_CLOEXEC, NETLINK_ROUTE)) < 0) {
         LOG(ERROR) << "Failed opening socket: " << strerror(errno);
-        return false;
+        return -int(eErrorCode::OPERATION_NOT_SUPPORTED);
```
Comment (tomereli) [inline framework/platform/bpl/uci/arp/monitor/arp_monitor.cpp:222]: Why is the cast needed here?

File: framework/platform/bpl/uci/arp/monitor/arp_monitor.cpp [hunk 7]
```diff
@@ -217,21 +219,21 @@ bool arp_monitor::start(std::string strIface)
...
     }
 
     int sndbuf = SOCKET_SEND_BUF_SIZE;
     if (setsockopt(m_fdMonSocket, SOL_SOCKET, SO_SNDBUF, &sndbuf, sizeof(sndbuf)) < 0) {
         LOG(ERROR) << "Failed setting SO_SNDBUF option: " << strerror(errno);
         stop();
-        return false;
+        return -1;
```
Comment (tomereli) [inline framework/platform/bpl/uci/arp/monitor/arp_monitor.cpp:229]: Why not return the enum as you do in other places?

**Review by tomereli — CHANGES_REQUESTED** [review]
First commit message typo:
opening a raw a raw socket will fail. --> opening raw socket will fail.

File: agent/src/beerocks/monitor/monitor_rssi.cpp (hunk 1 above)
Comment (morantr) [reply agent/src/beerocks/monitor/monitor_rssi.cpp:41]: nullptr assignment is not needed and not contributing anything. When the monitor calls to `mon_rssi.start()` and fails, it calls to `stop()` BUT it continue calling to `process()` on each work cycle, which uses these pointers.

File: agent/src/beerocks/monitor/monitor_rssi.cpp (hunk 2 above)
Comment (morantr) [reply agent/src/beerocks/monitor/monitor_rssi.cpp]: Yes, you right it should be `&&`. But no, it can't be inside arp_enabled() since it is two different things.

File: agent/src/beerocks/monitor/monitor_rssi.h (hunk 3 above)
Comment (morantr) [reply agent/src/beerocks/monitor/monitor_rssi.h:46]: No, the inactive arp is only for inactive arp, and not related to general arp.

File: agent/src/beerocks/slave/backhaul_manager/wan_monitor.cpp (hunk 5 above)
Comment (morantr) [reply agent/src/beerocks/slave/backhaul_manager/wan_monitor.cpp:36]: yes

File: framework/platform/bpl/uci/arp/monitor/arp_monitor.cpp (hunk 6 above)
Comment (morantr) [reply framework/platform/bpl/uci/arp/monitor/arp_monitor.cpp:222]: because the function returns `int`. 😕

File: framework/platform/bpl/uci/arp/monitor/arp_monitor.cpp (hunk 7 above)
Comment (morantr) [reply framework/platform/bpl/uci/arp/monitor/arp_monitor.cpp:229]: `-1` represents general failure without any specific error code.  I used the enum to separate permissions error form other general errors. Basically bpl_err.h doesn't have a general error enum, and I wasn't sure if it is a good idea to add one. Anyway `1` is already taken for `UNKNOWN` on bpl_err.h, unless you would like to replace the `-1` with `-int(eErrorCode::UNKNOWN)`.

File: agent/src/beerocks/monitor/monitor_thread.cpp (hunk 4 above)
Comment (morantr) [reply agent/src/beerocks/monitor/monitor_thread.cpp:278]: We don't need it here. A detailed error will be printed inside (as I wrote on the comment here).

File: agent/src/beerocks/monitor/monitor_rssi.cpp (hunk 2 above)
Comment (tomereli) [reply agent/src/beerocks/monitor/monitor_rssi.cpp]: Not sure I understand why it is two different things, but OK.

File: agent/src/beerocks/monitor/monitor_thread.cpp (hunk 4 above)
Comment (tomereli) [reply agent/src/beerocks/monitor/monitor_thread.cpp:278]: OK

**Review by tomereli — APPROVED** [review]
You answered all my comments so approving already.
